Reconstruct SELinux  mount label

jsafrane opened the pull request:
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:

Reconstruct SELinux mount option `-o context=XYZ` and add it to volume manager's Actual State of World (ASW) when kubelet starts. Kubelet then can see if a mounted volume in ASW has the right context that new Pods need and can unmount it and mount back to get the right SELinux context.

SELinux context for CSI volumes is determined from json file, for in-tree volumes it's taken directly from the mount table.

~In addition, there is a new e2e test that checks the reconstruction gets the right SELinux context and volume from an old pod is un-mounted when needed.~ E2e tests will be a separate PR.

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Added reconstruction of SELinux mount context after kubelet restart. Feature SELinuxMountReadWriteOncePod is now fully implemented and kubelet does not lose its cache of SELinux contexts after kubelet process restart.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
https://github.com/kubernetes/enhancements/pull/3548
<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/1710-selinux-relabeling
```

jsafrane committed:
Reconstruct SELinux mount option

When reconstructing volumes from disk after kubelet restart, reconstruct
also context=XYZ mount option and add it to the ActualStateOfWorld.

jsafrane committed:
Add SELinux disruptive test

jsafrane committed:
Update SELinux context log

SELinux context discovered from Pod is not final, it can be cleared when a
volume plugin does not support SELinux or the volume is not
ReadWriteOncePod. Update the existing log line + add a new one for easier
debugging.

jsafrane commented:
The e2e test needs some work, we can merge it after feature freeze.

jsafrane committed:
Add mock CSI driver test for SELinux mount

gnufied commented on pkg/kubelet/volumemanager/reconciler/reconstruct_new.go:
```diff
@@ -139,7 +142,7 @@ func (rc *reconciler) updateStatesNew(reconstructedVolumes map[v1.UniqueVolumeNa
 			}
-			err = rc.actualStateOfWorld.MarkDeviceAsUncertain(gvl.volumeName, gvl.devicePath, deviceMountPath, "")
+			err = rc.actualStateOfWorld.MarkDeviceAsUncertain(gvl.volumeName, gvl.devicePath, deviceMountPath, seLinuxMountContext)
 			if err != nil {
```
Should we have tests that verify if selinux information is recorded as expected?

jsafrane replied:
Unit test added

gnufied commented on pkg/volume/csi/csi_plugin.go:
```diff
@@ -455,22 +455,22 @@ func (p *csiPlugin) ConstructVolumeSpec(volumeName, mountPath string) (volume.Re
 
-	var spec *volume.Spec
+	ret := volume.ReconstructedVolume{
+		SELinuxMountContext: volData[volDataKey.seLinuxMountContext],
```
Should this be feature gated as others?

jsafrane replied:
Feature gate added

jsafrane commented:
Rebased, I'll address the feature gates + unit tests tomorrow.

jsafrane commented on pkg/kubelet/volumemanager/reconciler/reconstruct.go:
```diff
@@ -183,6 +183,7 @@ func (rc *reconciler) markVolumeState(volume *reconstructedVolume, volumeState o
 		VolumeSpec:          volume.volumeSpec,
 		VolumeMountState:    volumeState,
+		SELinuxMountContext: volume.seLinuxMountContext,
 	}
```
It does not, removed.

gnufied commented on pkg/kubelet/volumemanager/reconciler/reconstruct_new.go:
```diff
@@ -139,7 +142,7 @@ func (rc *reconciler) updateStatesNew(reconstructedVolumes map[v1.UniqueVolumeNa
+			err = rc.actualStateOfWorld.MarkDeviceAsUncertain(gvl.volumeName, gvl.devicePath, deviceMountPath, seLinuxMountContext)
 			if err != nil {
 				klog.ErrorS(err, "Could not mark device is uncertain to actual state of world", "volumeName", gvl.volumeName, "deviceMountPath", deviceMountPath)
 				continue
```
I am trying to verify(and I forgot some details), but is staging (ie. `MountDevice`) and Setup create different json files or they create same files? Are we consistently saving selinux context in both places? Is that feature gated?

jsafrane replied:
They create different jsons and I think SELinux is only in the SetUp one.

jsafrane replied:
Correction: it was in MountDevice, I added it to both + both feature-gated.

jsafrane committed:
Save SELinux context both in MountDevice and SetUp

And make it feature gated in both places.

gnufied commented:
/lgtm
